zkvm/src/constraints.rs

```diff
@@ -89,21 +92,61 @@ impl Constraint {
         .map_err(|e| VMError::R1CSError(e))
     }
 
-    fn flatten<CS: r1cs::RandomizedConstraintSystem>(self, cs: &mut CS) -> r1cs::LinearCombination {
+    fn flatten<CS: r1cs::RandomizedConstraintSystem>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(r1cs::LinearCombination, Option<Scalar>), r1cs::R1CSError> {
```

Comment:
this is not failable anymore?

Reply:
we need to be able to handle the errors from `allocate_multiplier`

Reply:
ouch. `r1cs::Prover` needs those assignments.

<img width="693" alt="image" src="https://user-images.githubusercontent.com/698/54305932-883a4d80-4585-11e9-924a-07c1f958c15f.png">